flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultBatchJobRecoveryHandler.java

```diff
@@ -791,18 +785,34 @@ private boolean isPartitionShouldBeReconciled(IntermediateResultPartition partit
                                 jobVertexId -> getExecutionJobVertex(jobVertexId).isInitialized());
 
         if (!allConsumersInitialized) {
-            return true;
+            return PartitionReservationStatus.RESERVE;
         }
 
-        // 3. If all downstream vertices are initialized, we need to check if the all partition
-        // consumers finished.
+        // 3. If all downstream vertices are finished, we need reserve the partitions. Otherwise, we
```

Comment:
all downstream vertices -> not all the consumer vertices

downstream vertices include vertices which do not directly consume the partition.